File: cosmos-api/app/es/tid/cosmos/api/controllers/info/info.scala
/*
 * Copyright (c) 2013-2014 Telefónica Investigación y Desarrollo S.A.U.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package es.tid.cosmos.api.controllers.info

import play.api.libs.json.{JsString, JsValue, Json, Writes}

import es.tid.cosmos.servicemanager.clusters.ClusterId

case class ResourcesInfo(
  groupConsumption: Int,
  individualConsumption: Int,
  available: Int,
  availableForGroup: Int,
  availableForUser: Int
)

object ResourcesInfo {
  implicit val resourcesInfoWrites = Json.writes[ResourcesInfo]
}

case class ClustersInfo(owned: Seq[ClusterId], accessible: Seq[ClusterId])

object ClustersInfo {

  implicit object ClusterIdWrites extends Writes[ClusterId] {
    override def writes(clusterId: ClusterId) = JsString(clusterId.id)
  }

  implicit val clustersInfoWrites = Json.writes[ClustersInfo]
}

case class GroupInfo(
  name: String,
  guaranteedQuota: Option[Int]
)

object GroupInfo {
  implicit object GroupInfoWrites extends Writes[GroupInfo] {
    override def writes(groupInfo: GroupInfo) =
      Json.obj("name" -> groupInfo.name) ++ quotaField(groupInfo.guaranteedQuota)

    private def quotaField(quotaOption: Option[Int]) = (
      for (quota <- quotaOption) yield Json.obj("guaranteedQuota" -> quota)
    ).getOrElse(Json.obj())
  }
}

/** General-purpose information about a platform user. */
case class Info(
  profileId: Long,
  handle: String,
  individualQuota: Option[Int],
  group: GroupInfo,
  clusters: ClustersInfo,
  resources: ResourcesInfo
)

object Info {
  implicit val infoWrites = Json.writes[Info]
}

File: cosmos-cli/cosmos/cli/info.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2013-2014 Telefónica Investigación y Desarrollo S.A.U.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""General info command"""

import json

from cosmos.cli.config import with_config
from cosmos.common.cosmos_requests import CosmosRequests
from cosmos.common.exceptions import ResponseError
from cosmos.common.routes import Routes


def info_command(args, config):
    response = CosmosRequests(config.credentials).get(Routes(config.api_url).info)
    if response.status_code != 200:
        raise ResponseError("Cannot get general information", response)
    print json.dumps(response.json(), sort_keys=True, indent=4)
    return 0


def add_info_command(subcommands):
    parser = subcommands.add_parser(
        "info", help="Show general information about the platform")
    parser.set_defaults(func=with_config(info_command))
